sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -163,81 +192,56 @@ public void forceFlush() {
               Labels.of("spanProcessorType", BatchSpanProcessor.class.getSimpleName()));
     }
 
-    private static final BoundLongCounter droppedSpans;
-
     private final ExecutorService executorService =
         Executors.newSingleThreadExecutor(new DaemonThreadFactory(EXPORTER_THREAD_NAME));
-
-    private static final Logger logger = Logger.getLogger(Worker.class.getName());
     private final SpanExporter spanExporter;
-    private final long scheduleDelayMillis;
     private final int maxQueueSize;
     private final int maxExportBatchSize;
     private final int halfMaxQueueSize;
-    private final Object monitor = new Object();
     private final int exporterTimeoutMillis;
-
-    @GuardedBy("monitor")
-    private final List<ReadableSpan> spansList;
+    private final BlockingQueue<ReadableSpan> spansList;
 
     private Worker(
         SpanExporter spanExporter,
-        long scheduleDelayMillis,
         int maxQueueSize,
         int maxExportBatchSize,
         int exporterTimeoutMillis) {
       this.spanExporter = spanExporter;
-      this.scheduleDelayMillis = scheduleDelayMillis;
       this.maxQueueSize = maxQueueSize;
       this.halfMaxQueueSize = maxQueueSize >> 1;
       this.maxExportBatchSize = maxExportBatchSize;
-      this.spansList = new ArrayList<>(maxQueueSize);
+      this.spansList = new ArrayBlockingQueue<>(maxQueueSize);
       this.exporterTimeoutMillis = exporterTimeoutMillis;
     }
 
+    private static List<SpanData> createSpanDataForExport(
+        List<ReadableSpan> spanList, int startIndex, int numberToTake) {
```

Comment:
Fix `numberToTake` to `endIndex`

Reply:
I think this is the number to take if you look how it's used in the exportBatches method.

Reply:
Really? We iterate from startIndex to this number in the loop

`for (int i = startIndex; i < numberToTake; i++) {`

If it was actually the number to take, it'd be 

`for (int i = startIndex; i < i + numberToTake; i++) {`

wouldn't it?

Reply:
yikes! This is a really bad bug. Thanks for the sharp eye on that! Good thing no one has been trying to send too many spans with this thing yet!

Reply:
ok, it wasn't really all that bad a bug, but @anuraaga is correct. The parameter is misnamed, both here and in the caller. This has been fixed on the main branch with #1400 , so if you could fix it here as well, that would be good.